app/bundles/LeadBundle/Model/LeadModel.php

```diff
@@ -1958,6 +1960,113 @@ public function setUtmTags(Lead $lead, UtmTag $utmTags)
     }
 
     /**
+     * Add leads UTM tags via API.
+     *
+     * @param Lead  $lead
+     * @param array $query
+     */
+    public function addUTMTags(Lead $lead, $query)
+    {
+        // available fields and it's setter.
+        $fields = ['utm_campaign' => 'UtmCampaign',
+                   'utm_source'   => 'UtmSource',
+                   'utm_medium'   => 'UtmMedium',
+                   'utm_content'  => 'UtmContent',
+                   'utm_term'     => 'UtmTerm',
+                   'useragent'    => 'UserAgent',
+                   'url'          => 'Url',
+                   'referer'      => 'Referer',
+                   'query'        => 'Query',
+                   'remotehost'   => 'RemoteHost',
+                   'lastActive'   => 'DateAdded',
+        ];
+
+        // 'query' must be an array
+        if (isset($query['query']) && !is_array($query['query'])) {
+            parse_str($query['query'], $queryResult);
+            if (!empty($queryResult)) {
+                $query['query'] = $queryResult;
+            } else {
+                // Something wrong with it
+                unset($query['query']);
+            }
+        }
+
+        // see if active date set, so we can use it
+        $updateLastActive = false;
+        $lastActive       = new \DateTime();
+        // should be: yyyy-mm-ddT00:00:00+00:00
+        if (isset($query['lastActive'])) {
+            $lastActive       = new \DateTime($query['lastActive']);
+            $updateLastActive = true;
+        }
+        $query['lastActive'] = $lastActive;
+
+        // New utmTag
+        $utmTags = new UtmTag();
+
+        // cycle through calling appropriate setter
+        foreach ($fields as $q => $prop) {
+            if (isset($query[$q])) {
+                $setter = 'set'.$prop;
+                $utmTags->$setter($query[$q]);
+            }
+        }
```

Comment:
I cannot find any place where `createFromArray()` is used.

What my concern is that if we define what methods the UtmTags class have in another place in an array, it will cause issues sooner or later when the UtmTags class changes.

I was discussing this with @alanhartless and he came up with 2 options:
1. Php is case insensitive for methods. We can InputHelper::alphanum the parameter, and prefix it with set, then do a method_exists, it'll work by looping over query rather than a hard coded array.
2. Use a public method within the UtmTags class to get the properties or setters available so that it's dynamically generated rather than hard coded.

I like the second option better.